fabric-registry-sync-v0/src/main/java/net/fabricmc/fabric/impl/registry/sync/packet/DirectRegistryPacketSerializer.java

```diff
@@ -0,0 +1,175 @@
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package net.fabricmc.fabric.impl.registry.sync.packet;
+
+import java.util.ArrayList;
+import java.util.Comparator;
+import java.util.Iterator;
+import java.util.LinkedHashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.stream.Collectors;
+
+import it.unimi.dsi.fastutil.objects.Object2IntLinkedOpenHashMap;
+import it.unimi.dsi.fastutil.objects.Object2IntMap;
+import org.jetbrains.annotations.Nullable;
+
+import net.minecraft.network.PacketByteBuf;
+import net.minecraft.util.Identifier;
+
+/**
+ * A more optimized method to sync registry ids to client.
+ * Produce smaller packet than old {@link NbtRegistryPacketSerializer nbt-based} method.
+ *
+ * <p>This method optimize the packet in multiple way:
+ * <ul>
+ *     <li>Directly write into the buffer instead of using an nbt;</li>
+ *     <li>Group all {@link Identifier} with same namespace together and only send those unique namespaces once for each group;</li>
+ *     <li>Group adjacent rawIds together and only send the difference of the first rawId and the last rawId of the bulk before.
+ *     This is based on the assumption that mods generally register all of their object at once,
+ *     therefore making the rawIds somewhat densely packed.</li>
+ * </ul>
+ */
+public class DirectRegistryPacketSerializer implements RegistryPacketSerializer {
+	public static final Identifier ID = new Identifier("fabric", "registry/sync/direct");
+
+	DirectRegistryPacketSerializer() {
+	}
+
+	@Override
+	public Identifier getPacketId() {
+		return ID;
+	}
+
+	@Override
+	public void writeBuffer(PacketByteBuf buf, Map<Identifier, Object2IntMap<Identifier>> map) {
+		// Group registry ids with same namespace
+		Map<String, List<Identifier>> regNamespaceGroups = map.keySet().stream()
+				.collect(Collectors.groupingBy(Identifier::getNamespace));
+
+		buf.writeVarInt(regNamespaceGroups.size());
+
+		regNamespaceGroups.forEach((regNamespace, regIds) -> {
+			buf.writeString(optimizeNamespace(regNamespace));
+			buf.writeVarInt(regIds.size());
+
+			for (Identifier regId : regIds) {
+				buf.writeString(regId.getPath());
+
+				Object2IntMap<Identifier> idMap = map.get(regId);
+
+				// Sort object ids by raw id and group them by namespace
+				Map<String, List<Object2IntMap.Entry<Identifier>>> idNamespaceGroups = idMap.object2IntEntrySet().stream()
+						.sorted(Comparator.comparingInt(Object2IntMap.Entry::getIntValue))
+						.collect(Collectors.groupingBy(e -> e.getKey().getNamespace(), LinkedHashMap::new, Collectors.toList()));
```

Comment:
Upon further reading, `SimpleRegistry`, which is what all registry basically is, is using `rawIdToEntry` list as backing for its iterator, making it already sorted by the raw id. Should we just don't bother re-sorting it here? I already used linked map for the id map in `RegistrySyncManager#createAndPopulateRegistryMap` anyway.

Reply:
Even if it is the case that SimpleRegistry's iteration order is by ascending raw ID, I don't know if it should be considered part of its Iterable contract or just an implementation detail that may change in the future. We would also have to ensure that items continue to be inserted into the linked map in that order in case we ever want to change `createAndPopulateRegistryMap` or the surrounding logic that provides the parameter to `writeBuffer`. If we are going to rely on this, it should at least be documented.